### DISCUSSION TITLE
Is there any way to prevent infinite growth except from timeouts?

### DISCUSSION THREAD

**Discussion by Strepto** [ZiggyCreatures/FusionCache#107] [Q&A] [answered]
Hey! Great library and impressive docs! But I did not find an answer to the question in the title:

Is there any way to prevent infinite growth in the MemoryCache except from using timeouts?

The use-case is for a API for a read-only SQLite-database cache layer where the query responses are idempotent (the response for a query never changes). The SQLite database files are replaced daily, so a cache time of ~24 hours is preferred, but if memory is spent the oldest request should be removed from memory.

There are potentially "infinite" queries that can be performed, and I want to trigger a cleanup in the style of either LRU (Least Recently Used) cache or FIFO, to clean X items (or size) cache to avoid infinite memory usage.

Does this library have any way to handle the cache size?

The dotnet `MemoryCache` has a concept of Size. Size gets problematic when using lazy, as the size is not determined before the request is finished, so I'm not sure if this can help with that.

**Comment by jodydonetti** [comment, marked as answer]
> Hey! Great library and impressive docs!

Thanks! I'm glad you are liking it 🙂

> But I did not find an answer to the question in the title:
> Is there any way to prevent infinite growth in the MemoryCache except from using timeouts?
> [...]
> The use-case is for a API for a read-only SQLite-database cache layer where the query responses are idempotent (the response for a query never changes). The SQLite database files are replaced daily, so a cache time of ~24 hours is preferred, but if memory is spent the oldest request should be removed from memory.

Interesting use-case, I never seen something like this! Would you be willing to explain more?
Like, is SQLite used as a 2nd level, did I get this right? And is the SQLite file being swapped directly, on-disk?

> There are potentially "infinite" queries that can be performed, and I want to trigger a cleanup in the style of either LRU (Least Recently Used) cache or FIFO, to clean X items (or size) cache to avoid infinite memory usage.
> 
> Does this library have any way to handle the cache size?

This library (currently) works on top of the common abstractions in .NET (`IMemoryCache` + `IDistributedCache`), so all the functionality available is built on top of those.

> The dotnet `MemoryCache` has a concept of Size. Size gets problematic when using lazy, as the size is not determined before the request is finished, so I'm not sure if this can help with that.

Exactly, I was about to suggest that! The limitation you are talking about though can be bypassed thanks to [Adaptive Caching](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AdaptiveCaching.md). Basically **during** the execution of the factory, when you just got back your data from your data source but before returning it, you can change some of the options like, in this case, the `Size`.

Hope this helps, let me know!

**Reply by Strepto** [reply]
Great answer. I did not notice the use of Size as an alternative in the Adaptive Caching implementation, but it seems like this can give me what I need.

The SQLite files are metadata mapped from a "View Model" of 3D Model. The 3D viewmodel and database needs to be in sync but will never change. 3D models are updated at regular intervals, so we then add a new SQLite database file and mark the old as ready for deletion in x days.

**Reply by jodydonetti** [reply]
> Great answer. I did not notice the use of Size as an alternative in the Adaptive Caching implementation, but it seems like this can give me what I need.

Nice! Let me know if it works for you.

> The SQLite files are metadata mapped from a "View Model" of 3D Model. The 3D viewmodel and database needs to be in sync but will never change. 3D models are updated at regular intervals, so we then add a new SQLite database file and mark the old as ready for deletion in x days.

Oh my 😍, I cannot begin to describe you how cool it is to read something like this as a scenario where FusionCache is being used!
Sooner or later I think I'll ask the community about how they are using it, in which scenarios, with which options and so on. I'd really like to know more.

Thanks for sharing 🙏
